From 37bfd5f970fd254b7cef5d3f6890c6da91aefd55 Mon Sep 17 00:00:00 2001
From: Robert Wiegand <qa2113@motorola.com>
Date: Tue, 28 Apr 2015 11:38:33 -0500
Subject: [PATCH 126/959] IKLOCSEN-2191 drivers/misc: stml0xx: Lux calculation
 devtree settings

Get the new lux calculation settings from the device tree and
pass them to the sensor hub.
This will allow the lux calculations to be customized on a
per product basis.

Change-Id: Id9d3c9598bccb6ca8d26a93f403568a47bcc387b
Signed-off-by: Robert Wiegand <qa2113@motorola.com>
Reviewed-on: http://gerrit.mot.com/741990
SME-Granted: SME Approvals Granted
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Jeffrey Carlyle <jeff.carlyle@motorola.com>
Reviewed-by: Jason Jae Hyun Koh <jasonkoh@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
Reviewed-by: Gabriel Burca <Gabriel.Burca@motorola.com>
---
 drivers/misc/stml0xx/stml0xx_core.c  | 18 ++++++++++++++++++
 drivers/misc/stml0xx/stml0xx_reset.c | 14 +++++++++++++-
 include/linux/stml0xx.h              |  6 ++++++
 3 files changed, 37 insertions(+), 1 deletion(-)

diff --git a/drivers/misc/stml0xx/stml0xx_core.c b/drivers/misc/stml0xx/stml0xx_core.c
index f909f2b9106..da002134a6d 100644
--- a/drivers/misc/stml0xx/stml0xx_core.c
+++ b/drivers/misc/stml0xx/stml0xx_core.c
@@ -412,6 +412,24 @@ static struct stml0xx_platform_data *stml0xx_of_init(struct spi_device *spi)
 			     &pdata->ct406_pulse_count);
 	of_property_read_u32(np, "ct406_prox_gain",
 			     &pdata->ct406_prox_gain);
+	pdata->ct406_als_lux1_c0_mult = 0x294;
+	pdata->ct406_als_lux1_c1_mult = 0x55A;
+	pdata->ct406_als_lux1_div = 0x64;
+	pdata->ct406_als_lux2_c0_mult = 0xDA;
+	pdata->ct406_als_lux2_c1_mult = 0x186;
+	pdata->ct406_als_lux2_div = 0x64;
+	of_property_read_u32(np, "ct406_als_lux1_c0_mult",
+			     &pdata->ct406_als_lux1_c0_mult);
+	of_property_read_u32(np, "ct406_als_lux1_c1_mult",
+			     &pdata->ct406_als_lux1_c1_mult);
+	of_property_read_u32(np, "ct406_als_lux1_div",
+			     &pdata->ct406_als_lux1_div);
+	of_property_read_u32(np, "ct406_als_lux2_c0_mult",
+			     &pdata->ct406_als_lux2_c0_mult);
+	of_property_read_u32(np, "ct406_als_lux2_c1_mult",
+			     &pdata->ct406_als_lux2_c1_mult);
+	of_property_read_u32(np, "ct406_als_lux2_div",
+			     &pdata->ct406_als_lux2_div);
 
 	pdata->dsp_iface_enable = 0;
 	of_property_read_u32(np, "dsp_iface_enable",
diff --git a/drivers/misc/stml0xx/stml0xx_reset.c b/drivers/misc/stml0xx/stml0xx_reset.c
index 2ae2abf3662..a9b919401ab 100644
--- a/drivers/misc/stml0xx/stml0xx_reset.c
+++ b/drivers/misc/stml0xx/stml0xx_reset.c
@@ -181,8 +181,20 @@ void stml0xx_initialize_work_func(struct work_struct *work)
 	buf[5] = pdata->ct406_recalibrate_threshold & 0xff;
 	buf[6] = pdata->ct406_pulse_count & 0xff;
 	buf[7] = pdata->ct406_prox_gain & 0xff;
+	buf[8] = (pdata->ct406_als_lux1_c0_mult >> 8) & 0xff;
+	buf[9] = pdata->ct406_als_lux1_c0_mult & 0xff;
+	buf[10] = (pdata->ct406_als_lux1_c1_mult >> 8) & 0xff;
+	buf[11] = pdata->ct406_als_lux1_c1_mult & 0xff;
+	buf[12] = (pdata->ct406_als_lux1_div >> 8) & 0xff;
+	buf[13] = pdata->ct406_als_lux1_div & 0xff;
+	buf[14] = (pdata->ct406_als_lux2_c0_mult >> 8) & 0xff;
+	buf[15] = pdata->ct406_als_lux2_c0_mult & 0xff;
+	buf[16] = (pdata->ct406_als_lux2_c1_mult >> 8) & 0xff;
+	buf[17] = pdata->ct406_als_lux2_c1_mult & 0xff;
+	buf[18] = (pdata->ct406_als_lux2_div >> 8) & 0xff;
+	buf[19] = pdata->ct406_als_lux2_div & 0xff;
 	err = stml0xx_spi_send_write_reg_reset(PROX_SETTINGS, buf,
-			8, RESET_NOT_ALLOWED);
+			20, RESET_NOT_ALLOWED);
 	if (err < 0) {
 		dev_err(&ps_stml0xx->spi->dev,
 			"unable to write proximity settings %d", err);
diff --git a/include/linux/stml0xx.h b/include/linux/stml0xx.h
index 70e3f765380..7fd2ec6aacb 100644
--- a/include/linux/stml0xx.h
+++ b/include/linux/stml0xx.h
@@ -623,6 +623,12 @@ struct stml0xx_platform_data {
 	int ct406_recalibrate_threshold;
 	int ct406_pulse_count;
 	int ct406_prox_gain;
+	int ct406_als_lux1_c0_mult;
+	int ct406_als_lux1_c1_mult;
+	int ct406_als_lux1_div;
+	int ct406_als_lux2_c0_mult;
+	int ct406_als_lux2_c1_mult;
+	int ct406_als_lux2_div;
 	int dsp_iface_enable;
 	int headset_detect_enable;
 	int headset_hw_version;
-- 
2.11.0

